pkg/cmd/admin/node/node.go

```diff
@@ -0,0 +1,399 @@
+	}
+
+	selector, err := labels.Parse(cfg.PodSelector.Value())
+	if err != nil {
+		return err
+	}
+
+	pods, err := cfg.Kclient.Pods(kapi.NamespaceAll).List(selector, fields.Everything())
+	if err != nil {
+		return err
+	}
+	rcs, err := cfg.Kclient.ReplicationControllers(kapi.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	printerWithHeaders, printerNoHeaders, err := cfg.getPrinters("pod", nil)
+	if err != nil {
+		return err
+	}
+
+	errList := []error{}
+	firstPod := true
+	// grace = 0 implies delete the pod immediately
+	grace := int64(0)
+	deleteOptions := &kapi.DeleteOptions{GracePeriodSeconds: &grace}
+
+	for _, pod := range pods.Items {
+		if pod.Spec.Host != node.ObjectMeta.Name {
+			continue
+		}
+
+		foundrc := false
+		for _, rc := range rcs.Items {
+			selector := labels.SelectorFromSet(rc.Spec.Selector)
+			if selector.Matches(labels.Set(pod.Labels)) {
+				foundrc = true
+				break
+			}
+		}
+
+		if firstPod {
+			fmt.Fprintln(cfg.Writer, "\nMigrating these pods on node: ", node.ObjectMeta.Name, "\n")
+			firstPod = false
+			printerWithHeaders.PrintObj(&pod, cfg.Writer)
+		} else {
+			printerNoHeaders.PrintObj(&pod, cfg.Writer)
+		}
+
+		if err := cfg.Kclient.Pods(pod.Namespace).Delete(pod.Name, deleteOptions); err != nil {
+			glog.Errorf("Unable to delete a pod: %+v, error: %v", pod, err)
+			errList = append(errList, err)
+			continue
+		}
+		if !foundrc {
+			// For pods without replication controller, pod.Spec.Host set by user will not be honored.
+			// more details: https://github.com/GoogleCloudPlatform/kubernetes/issues/8007
+			pod.Spec.Host = ""
+			pod.ObjectMeta.ResourceVersion = ""
+			pod.Status = kapi.PodStatus{}
+			_, err := cfg.Kclient.Pods(pod.Namespace).Create(&pod)
+			if err != nil {
+				glog.Errorf("Unable to create a pod: %+v, error: %v", pod, err)
+				errList = append(errList, err)
+				continue
+			}
+		}
+	}
+	if len(errList) != 0 {
+		return fmt.Errorf("%v", errList)
+	}
+	return nil
+}
+
+func (cfg *NodeConfig) RunListPods(node *kapi.Node) error {
+	selector, err := labels.Parse(cfg.PodSelector.Value())
+	if err != nil {
+		return err
+	}
+
+	pods, err := cfg.Kclient.Pods(kapi.NamespaceAll).List(selector, fields.Everything())
+	if err != nil {
+		return err
+	}
+
+	printerWithHeaders, printerNoHeaders, err := cfg.getPrinters("pod", nil)
+	if err != nil {
+		return err
+	}
+	firstPod := true
+
+	for _, pod := range pods.Items {
+		if pod.Spec.Host != node.ObjectMeta.Name {
+			continue
+		}
+
+		if firstPod {
+			fmt.Fprintln(cfg.Writer, "\nListing matched pods on node: ", node.ObjectMeta.Name, "\n")
+			printerWithHeaders.PrintObj(&pod, cfg.Writer)
+			firstPod = false
+		} else {
+			printerNoHeaders.PrintObj(&pod, cfg.Writer)
+		}
+	}
+	return err
+}
+
+func (cfg *NodeConfig) getPrinters(resource string, obj runtime.Object) (kubectl.ResourcePrinter, kubectl.ResourcePrinter, error) {
```

Comment:
Looks like three methods.  `getPrinters(version, kind)`, `getPrintersByResource`, and `getPrintersByObject` with delegation between them.  It's confusing to deal with mutually exclusive, "sometimes" method parameters.